### PR TITLE
Build glog with cmake as a bazel foreign rule

### PR DESCRIPTION
**Pull request by YuzhouGuo**
To solve the dump issue:
- https://github.com/cyfitech/cris-core/pull/96

### CODE REVIEW

**Review by chenhao94 — COMMENTED** [review]
pcre是啥？我们不是在改glog的编译么

**Comment by YuzhouGuo** [comment]
Seems like there're some extra packages here, also part of the addition can be replaced by the local path.

[This](https://docs.aspect.build/bazelbuild/rules_foreign_cc/0.3.0/docs/cmake.html) is what I referenced, researching now.

**Comment by chenhao94** [comment]
那只是个例子，我们代码里也有同样的例子，比如编译simdjson

**Comment by chenhao94** [comment]
check https://github.com/cyfitech/cris-core/blob/master/bazel/deps.bzl

And also you need to make sure stacktraces are printed after patching

**Comment by YuzhouGuo** [comment]
Yeah this is what I'm doing now, seems to be on the right track and this is the only file to be modified.

**Comment by YuzhouGuo** [comment]
'''
def cris_deps_glog(prefix = "."):
    if not native.existing_rule("glog"):
        native.new_local_repository(
            name = "com_github_google_glog",
            path = prefix + "/external/glog",
            build_file_content = """
load("@rules_foreign_cc//foreign_cc:defs.bzl", "cmake")

filegroup(
    name = "all_content",
    srcs = glob(["**"]),
    visibility = ["//visibility:public"],
)

cmake(
    name = "glog",
    lib_source = ":all_content",
    generate_args = [
        "-G Ninja",
    ],
    visibility = ["//visibility:public"],
)
            """,
        )
'''
This is what I did, and seems like the .a file wasn't exported to the correct path, trying to figure this out for a while

Terminal output from the build bash script:
'''
yuzhou@yuzhou-System-Product-Name:~/YuzhouStuff/cris/external/cris-xchg/external/cris-base/external/cris-core$ bazel build //tests:message_test
INFO: Analyzed target //tests:message_test (63 packages loaded, 8721 targets configured).
INFO: Found 1 target...
ERROR: /home/yuzhou/.cache/bazel/_bazel_yuzhou/2f2ff29c37d25cd8be7c3bd95760a32c/external/com_github_google_glog/BUILD.bazel:10:6: output 'external/com_github_google_glog/glog/lib/glog.a' was not created
ERROR: /home/yuzhou/.cache/bazel/_bazel_yuzhou/2f2ff29c37d25cd8be7c3bd95760a32c/external/com_github_google_glog/BUILD.bazel:10:6: Foreign Cc - CMake: Building glog failed: not all outputs were created or valid
Target //tests:message_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
ERROR: /home/yuzhou/YuzhouStuff/cris/external/cris-xchg/external/cris-base/external/cris-core/tests/BUILD:12:14 Linking tests/message_test failed: not all outputs were created or valid
INFO: Elapsed time: 6.051s, Critical Path: 5.79s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
FAILED: Build did NOT complete successfully
'''

**Comment by chenhao94** [comment]
Bazel will look for the "default" output, which should be "name.a/so". If your output has a different filename, you need to specify it.

**Comment by chenhao94** [comment]
https://bazelbuild.github.io/rules_foreign_cc/0.5.1/cmake.html#cmake-out_static_libs

**Comment by YuzhouGuo** [comment]
Working on two issues now:

- For some reason, the output was generated to the `copy_glog` directory instead of `glog`, guess it's because bazel did same work before?
- Outputting a `.so` file instead of `.a` file

**Comment by chenhao94** [comment]
> - For some reason, the output was generated to the `copy_glog` directory instead of `glog`, guess it's because bazel did same work before?

Bazel will copy the source code to a "sandbox" and start from there, but that's the implementation detail of Bazel and usually you don't need to care about. Just focus on what cmake does, where is the library. If you don't know, you can try cmake yourself.

> - Outputting a `.so` file instead of `.a` file

`.so` is dynamic/shared library while `.a` is static library. Usually it is configurable by cmake options. We prefer static linking.

**Comment by YuzhouGuo** [comment]
Got stuck with option 3 (Turns out the very first change didn't work as well, forget to delete a manual flag value set).
Done with option 4, meaning we took full control of the signal flags ourselves in `cr_signal.cc`. Additional flag options were copied from the google repo.

**Comment by YuzhouGuo** [comment]
Let me try with the cmake production myself to learn more details.

**Comment by YuzhouGuo** [comment]
I checked the cmake log, turns out it has nothing to do with the cmake output path/location, the cmake process did generate the correct directory and layers, but not the static lib at all. Latest educated guess is that there's dependency lacking.

**Comment by chenhao94** [comment]
> I checked the cmake log, turns out it has nothing to do with the cmake output path/location, the cmake process did generate the correct directory and layers, but not the static lib at all. Latest educated guess is that there's dependency lacking.

No, that is because you need to configure it. See https://github.com/google/glog/blob/b33e3bad4c46c8a6345525fd822af355e5ef9446/CMakeLists.txt#L37 . The default choice is shared library while we want the static library.

**Comment by YuzhouGuo** [comment]
I did that before as well, I configured 
```
        "BUILD_SHARED_LIBS": "OFF",
        "BUILD_STATIC_LIBS": "ON",
```
and it didn't work, same error message saying no `.a` file output.

Since gflag is the dependency of glog, I also tried cmake gflag first then glog, but seems no difference.

File: bazel/deps.bzl [hunk 1]
```diff
@@ -7,9 +7,32 @@ def cris_deps_gflags(prefix = "."):
...
+filegroup(
+    name = "all_content",
+    srcs = glob(["**"]),
+    visibility = ["//visibility:private"],
+)
+
+cmake(
+    name = "glog",
+    lib_source = ":all_content",
+    cache_entries = {
```
[the remark this answers is not in the document]
Comment (xkszltl) [reply bazel/deps.bzl:25]: @chenhao94 还是说bazel会自动加上？

File: bazel/deps.bzl [hunk 2]
```diff
@@ -7,9 +7,32 @@ def cris_deps_gflags(prefix = "."):
...
+        "CMAKE_C_FLAGS" : "-Wno-fuse-ld-path",
+        "CMAKE_CXX_FLAGS" : "-Wno-fuse-ld-path",
+        "BUILD_SHARED_LIBS": "OFF",
```
[the remark this answers is not in the document]
Comment (xkszltl) [reply bazel/deps.bzl:28]: @chenhao94 为啥prefer static？对于这个case我觉得都行，只是问问

[the remark this answers is not in the document]
Comment (YuzhouGuo) [reply bazel/deps.bzl:28]: not sure about this @chenhao94

**Comment by xkszltl** [comment]
image

这里应该用完整的URL，而不只是一个`#96`，URL建议放在新的一行做成item list，这样github会显示详细的状态

```
XXXXX
- https://github.com/cyfitech/cris-core/...
```

**Comment by xkszltl** [comment]
比如
image
image

File: bazel/deps.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenhao94) [reply bazel/deps.bzl:28]: - Bazel默认是静态链接，这样和其他cc_library一致
- release的时候不用操心这些动态链接库

这也不是很强的理由，所以只是preference不是requirement

File: bazel/deps.bzl (hunk 1 above)
[the remark this answers is not in the document]
Comment (chenhao94) [reply bazel/deps.bzl:25]: Release会根据bazel的compilation mode判断，我们是`-c opt`那就是release。

https://github.com/bazelbuild/rules_foreign_cc/blob/2c6262f8f487cd3481db27e2c509d9e6d30bfe53/foreign_cc/cmake.bzl#L209

`-g`不会，如果都需要我们可以干脆加bazelrc里

File: bazel/deps.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (xkszltl) [reply bazel/deps.bzl:28]: > release的时候不用操心这些动态链接库

FYI这个不一定成立，静态库有可能自己依赖动态库

File: bazel/deps.bzl (hunk 1 above)
[the remark this answers is not in the document]
Comment (xkszltl) [reply bazel/deps.bzl:25]: 可以加，但注意并不是所有project都支持ninja（我们会用到的应该基本都支持）

[the remark this answers is not in the document]
Comment (YuzhouGuo) [reply bazel/deps.bzl:25]: @chenhao94

File: bazel/deps.bzl (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenhao94) [reply bazel/deps.bzl:28]: 对，但底下依赖的库都一样，静态链接少依赖了当前这一层

**Pull request closed without merging**